contributors/design-proposals/pod-resource-management.md

```diff
@@ -1,114 +1,480 @@
+priority, the risk of an update being incorrect is minimized as the `kubelet`
+converges to a desired state.  Failure to set `cpu.shares` at the QoS level
+cgroup would result in `500m` of cpu for a Guaranteed pod to have different
+meaning than `500m` of cpu for a Burstable pod in the current hierarchy.  This
+is because the default `cpu.shares` value if unspecified is `1024` and `cpu.shares`
+are evaluated relative to sibling nodes in the cgroup hierarchy.  As a consequence,
+all of the Burstable pods under contention would have a relative priority of 1 cpu
+unless updated dynamically to capture the sum of requests.  For this reason,
+we will always set `cpu.shares` for the QoS level sandboxes
+by default as part of roll-out for this feature.
+
+### QoS level memory allocation
+
+By default, no memory limits are applied to the BestEffort
+and Burstable QoS level cgroups unless a `--qos-reserve-requests` value
+is specified for memory.
+
+The heuristic that is applied is as follows for each QoS level sandbox:
+
+```
+ROOT/Burstable/memory.limit_in_bytes = 
+    Node.Allocatable - {(summation of memory requests of `Guaranteed` pods)*(reservePercent / 100)}
+ROOT/BestEffort/memory.limit_in_bytes = 
+    Node.Allocatable - {(summation of memory requests of all `Guaranteed` and `Burstable` pods)*(reservePercent / 100)}
+```
+
+A value of `--experimental-qos-reserved=memory=100%` will cause the
+`kubelet` to adjust the Burstable and BestEffort cgroups from consuming memory
+that was requested by a higher QoS class. This increases the risk
+of inducing OOM on BestEffort and Burstable workloads in favor of increasing
+memory resource guarantees for Guaranteed and Burstable workloads.  A value of
+`--experimental-qos-reserved=memory=0%` will allow a Burstable
+and BestEffort QoS sandbox to consume up to the full node allocatable amount if
+available, but increases the risk that a Guaranteed workload will not have
+access to requested memory.
+
+Since memory is an incompressible resource, it is possible that a QoS
+level cgroup sandbox may not be able to reduce memory usage below the
+value specified in the heuristic described earlier during pod admission
+and pod termination.
+
+As a result, the `kubelet` runs a periodic thread to attempt to converge
+to this desired state from the above heuristic.  If unreclaimable memory
+usage has exceeded the desired limit for the sandbox, the `kubelet` will
+attempt to set the effective limit near the current usage to put pressure
+on the QoS cgroup sandbox and prevent further consumption.
+
+The `kubelet` will not wait for the QoS cgroup memory limit to converge
+to the desired state prior to execution of the pod, but it will always
+attempt to cap the existing usage of QoS cgroup sandboxes in lower tiers.
+This does mean that the new pod could induce an OOM event at the `ROOT`
+cgroup, but ideally per our QoS design, the oom_killer targets a pod
```

Comment:
As a next step, we need to enable evictions at the QoS levels too to facilitate QoS enforcement.